README.md
# react-availability

Availability search for the web in React.

## Installation

Start by downloading and building the project when necessary. The following
commands will do the job on most Debian based Linux distributions.

```bash
git clone https://github.com/marcbperez/react-availability
cd react-availability
sudo ./gradlew
```

## Usage

Run the following script to start a development server. After this the
application will be available at `http://127.0.0.1:3000`.

```bash
npm start
```

To get and run a production build use the script below. The contents will be
placed inside `build`. Tests can be executed via docker-compose (described in
the section below) or executed manually with NPM and Jest.

```bash
npm test
npm build
npm run build
```

## Testing

Test checks are executed automatically every time the project is built. Builds
can be done remotely or continuously on a development context. For continuous
integration and development use docker-compose. This is recommended to keep the
system clean while the project is built every time the sources change.

```bash
sudo docker-compose up
```

For continuous integration and development without any dependencies use the
Gradle wrapper. This is the best option if the wrapper is available and the
Docker context is not valid. For a full list of tasks, see
`sudo ./gradlew tasks --all`. For a CI cycle use `sudo ./gradlew --continuous`.

For continuous integration and development without Docker or the project wrapper
use Gradle directly. This will create the wrapper in case it is not present.
Similar to the above, for a CI cycle use `sudo gradle --continuous`. Gradle
3.4.1 is required for this to work. Plain Docker is also available for remote
integration tasks and alike. Build the image with `sudo docker build .` and run
a new container with it. Information on how to install Docker and docker-compose
can be found in their [official page][install-docker-compose]. A similar
installation guide is available [for Gradle][install-gradle].

## Troubleshooting

The [issue tracker][issue-tracker] intends to manage and compile bugs,
enhancements, proposals and tasks. Reading through its material or reporting to
its contributors via the platform is strongly recommended.

## Contributing

This project adheres to [Semantic Versioning][semver] and to certain syntax
conventions defined in [.editorconfig][editorconfig]. To get a list of changes
refer to the [CHANGELOG][changelog]. Only branches prefixed by *feature-*,
*hotfix-*, or *release-* will be considered:

  - Fork the project.
  - Create your new branch: `git checkout -b feature-my-feature develop`
  - Commit your changes: `git commit -am 'Added my new feature.'`
  - Push the branch: `git push origin feature-my-feature`
  - Submit a pull request.

## Credits

This project is created by [marcbperez][author] and maintained by its
[author][author] and contributors.

## License

This project is licensed under the [Apache License Version 2.0][license].

[author]: https://marcbperez.github.io
[issue-tracker]: https://github.com/marcbperez/react-availability/issues
[editorconfig]: .editorconfig
[changelog]: CHANGELOG.md
[license]: LICENSE
[semver]: http://semver.org
[install-docker-compose]: https://docs.docker.com/compose/install/
[install-gradle]: https://gradle.org/install


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. This
project adheres to [Semantic Versioning](http://semver.org).

## Unreleased

## 0.4.0 - 2018-04-17

### Changed

  - Project owner's name.

## 0.3.2 - 2018-04-09

  - Corrections in README and main application file.

## 0.3.1 - 2018-04-08

### Changed

  - Project name and README information.
  - Service class to a cleaner version.
  - Fetch for cross-fetch package.
  - Gradle wrapper for this project.

## 0.3.0 - 2018-04-07

### Added

  - Package configuration file.
  - Public folder with template and favicon.
  - Application index and service.
  - Application, day and slot components.

## 0.2.0 - 2018-04-06

### Added

  - Gradle build script and wrapper.
  - Dockerfile and docker-compose manifest.

## 0.1.0 - 2018-04-05

### Added

  - A README which covers project description and usage.
  - This CHANGELOG file to document project changes and versions.
  - An EditorConfig to suggest syntax and format standards.
  - A LICENSE file to define project ownership and rights.
  - A .gitignore list for untracked resources.


src/App/App.test.js
import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import Service from '../Service'

const fetch = jest.fn(() => new Promise(resolve => resolve()));
const e = {
  preventDefault: () => {},
  target: {
    type: 'text'
  }
};

it('renders without crashing', () => {
  const service = new Service('');
  const div = document.createElement('div');
  ReactDOM.render(<App service={service}/>, div);
});

it('searches for availability', () => {
  const service = new Service('');
  const app = new App({service: service});
  app.searchChange(e);
  app.searchSubmit(e);
  app.updateDays([{}]);
});


src/Day/Day.test.js
import React from 'react';
import ReactDOM from 'react-dom';
import Day from './Day';
import Service from '../Service'

it('renders without crashing', () => {
  const service = new Service('');
  const div = document.createElement('div');
  ReactDOM.render(
    <Day service={service} startTimes={[{ possible: true }]}/>, div
  );
});

it('shows fully booked', () => {
  const service = new Service('');
  const div = document.createElement('div');
  ReactDOM.render(
    <Day service={service} startTimes={[{ possible: false }]}/>, div
  );
});


src/Service.test.js
import Service from './Service';

const fetch = jest.fn(() => new Promise(resolve => resolve()));

it('gets availability', () => {
  const service = new Service('');
  service.getAvailability(function(data) {}, '2016-12-05', 2.5, 'EC1R 3BU');
});

it('creates a booking', () => {
  const service = new Service('');
  service.createBooking(
    function(data) {}, '2016-12-11', '08:00:00', '08:30:00', 2.5, 'ealdk1f9'
  );
});


src/Day/Day.js
import React, { Component } from 'react';
import Slot from '../Slot/Slot';
import './Day.css';

/**
 * Day component.
 * @extends Component
 */
class Day extends Component {
  /**
   * Render the component.
   */
  render() {
    const service = this.props.service;
    const day = this.props.day;
    const visitDuration = this.props.visitDuration;
    const propertyId = this.props.propertyId;

    /* Only consider the slots that are available. */
    const startTimes = this.props.startTimes.filter(
      starTime => starTime.possible
    );

    return (
      <div className='day'>
        <h3>{this.props.day}</h3>

        {!startTimes.length ? (
          <p className='status'>Fully booked</p>
        ) : (
          <p className='status'>{startTimes.length} slots available</p>
        )}

        {startTimes.map(function(startTime, key) {
          return (
            <Slot
              key={key}
              service={service}
              day={day}
              start={startTime.start}
              end={startTime.end}
              visitDuration={visitDuration}
              propertyId={propertyId}
            />
          );
        })}
      </div>
    );
  }
}

export default Day;


src/Service.js
import fetch from 'cross-fetch';

/**
 * Application service
 */
class Service {
  /**
   * Initialize the service.
   * @param {string} baseUrl - The base url of the service.
   */
  constructor(baseUrl) {
    this.baseUrl = baseUrl;
  }

  /**
   * Return an url parameter.
   * @param {string} name - The name of the parameter.
   * @param {string} value - The value of the parameter.
   */
  urlParameter(name, value) {
    return '&' + name + '=' + value;
  }

  /**
   * Get availability.
   * @param {function} callback - The callback to send data to.
   * @param {string} weekBeginning - The week beginning to query.
   * @param {decimal} visitDuration - The visit duration in hours.
   * @param {string} postCode - The postcode of the property.
   */
  getAvailability(callback, weekBeginning, visitDuration, postcode) {
    let url = this.baseUrl + '/availability/?';

    /* Add parameters to the url. */
    url += this.urlParameter('weekBeginning', weekBeginning);
    url += this.urlParameter('visitDuration', visitDuration);
    url += this.urlParameter('postcode', postcode);

    /* Prepare fetch parameters. */
    var init = {
      method: 'GET',
      cache: 'default',
    };

    /* Fetch data and send to callback. */
    fetch(url, init)
      .then(function(response) {
        return response.json();
      }).then(function(data) {
        callback(data);
      });
  }

  /**
   * Create a booking.
   * @param {function} callback - The callback to send data to.
   * @param {string} day - The day to book.
   * @param {string} startTime - The start time of the slot.
   * @param {string} endTime - The end time of the slot.
   * @param {decimal} visitDuration - The visit duration in hours.
   * @param {string} propertyId - The property identifier.
   */
  createBooking(callback, day, startTime, endTime, visitDuration, propertyId) {
    let url = this.baseUrl + '/book/';

    /* Prepare fetch headers. */
    var headers = {
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    }

    /* Define request body. */
    var body = {
      'day': day,
      'startTime': {
        'start': startTime,
        'end': endTime
      },
      'visitDuration': visitDuration,
      'propertyId': propertyId
    }

    /* Prepare fetch parameters. */
    var init = {
      method: 'POST',
      cache: 'default',
      headers: headers,
      body: JSON.stringify(body)
    };

    /* Fetch data and send to callback. */
    fetch(url, init)
      .then(function(response) {
        return response.json();
      }).then(function(data) {
        callback(data);
      });
  }
}

export default Service;


src/Slot/Slot.js
import React, { Component } from 'react';
import './Slot.css'

/**
 * Slot component.
 * @extends Component
 */
class Slot extends Component {
  /**
   * Initialize the component.
   * @param {Object} props - The component props.
   */
  constructor(props) {
    super(props);

    this.state = {
      cleaner: ''
    };

    this.book = this.book.bind(this);
    this.update = this.update.bind(this);
  }

  /**
   * Send a booking request to the service.
   * @param {Event} e - The default event.
   */
  book(e) {
    e.preventDefault();
    this.props.service.createBooking(
      this.update,
      this.props.day,
      this.props.start,
      this.props.end,
      this.props.visitDuration,
      this.props.propertyId
    );
  }

  /**
   * Update the slot after it has been booked.
   * @param {Object} data - The data to update from.
   */
  update(data) {
    if (data.success) {
      this.setState({
        cleaner: data.cleaner.name.slice()
      })
    }
  }

  /**
   * Render the component.
   */
  render() {
    return (
      <p className='slot'>
        {this.props.start} to {this.props.end}
        {this.state.cleaner.length ? (
          <span>Booked with {this.state.cleaner}</span>
        ) : (
          <a href="#" onClick={this.book}>Book</a>
        )}
      </p>
    );
  }
}

export default Slot;


src/App/App.js
import React, { Component } from 'react';
import Day from '../Day/Day';
import './App.css';

/**
 * Main application.
 * @extends Component
 */
class App extends Component {
  /**
   * Initialize the application.
   * @param {Object} props - The application props.
   */
  constructor(props) {
    super(props);

    this.state = {
      days: Array(0),
      weekBeginning: '2016-12-05',
      visitDuration: 2.5,
      postcode: 'EC1R 3BU',
      propertyId: 'ealdk1f9'
    };

    this.updateDays = this.updateDays.bind(this);
    this.searchChange = this.searchChange.bind(this);
    this.searchSubmit = this.searchSubmit.bind(this);

    this.props.service.getAvailability(
      this.updateDays,
      this.state.weekBeginning,
      this.state.visitDuration,
      this.state.postcode
    );
  }

  /**
   * Update days after a search.
   * @param {Object} data - The data to update from.
   */
  updateDays(data) {
    this.setState({
      days: data.slice()
    });
  }

  /**
   * Updates state values from search form.
   * @param {Event} e - The default event.
   */
  searchChange(e) {
    const target = e.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const name = target.name;

    this.setState({
      [name]: value
    });
  }

  /**
   * Submits the search form and updates the days.
   * @param {Event} e - The default event.
   */
  searchSubmit(e) {
    e.preventDefault();
    this.props.service.getAvailability(
      this.updateDays,
      this.state.weekBeginning,
      this.state.visitDuration,
      this.state.postcode
    );
  }

  /**
   * Render the main application.
   */
  render() {
    const service = this.props.service;
    const visitDuration = this.props.visitDuration;
    const propertyId = this.state.propertyId;

    return (
      <div>
        <div className="app-header">
          <h2>React Availability</h2>
          <form onSubmit={this.searchSubmit}>
            <div>
              <label>Week beginning</label>
              <input name="weekBeginning" type="date"
                value={this.state.weekBeginning}
                onChange={this.searchChange} />
            </div>
            <div>
              <label>Visit duration</label>
              <input name="visitDuration" type="number" step="0.5"
                value={this.state.visitDuration}
                onChange={this.searchChange} />
            </div>
            <div>
              <label>Postcode</label>
              <input name="postcode" type="text"
                value={this.state.postcode}
                onChange={this.searchChange} />
            </div>
            <input type="submit" value="Search" />
          </form>
        </div>
        <div className="app-content">
          {this.state.days.map(function(day, key) {
            return (
              <Day
                key={key}
                service={service}
                day={day.day}
                startTimes={day.startTimes}
                visitDuration={visitDuration}
                propertyId={propertyId}
              />
            );
          })}
        </div>
      </div>
    );
  }
}

export default App;
